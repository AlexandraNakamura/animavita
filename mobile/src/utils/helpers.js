import AsyncStorage from '@react-native-community/async-storage';

export const sanitizeUsername = (name) => {
  if (name && name.length > 8) {
    return `\n ${name}`;
  }

  return name;
};

export const isEmpty = array => array instanceof Array && array.length === 0;

export const getToken = async () => JSON.parse(await AsyncStorage.getItem('@animavita:authToken')) || {};
