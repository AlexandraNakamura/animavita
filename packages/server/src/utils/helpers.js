import AWS from 'aws-sdk';

require('dotenv').config();

const {
  S3_KEY, S3_SECRET, S3_BUCKET, S3_REGION
} = process.env;

export const getAnimalGender = (gender) => {
  const genders = [
    {
      key: 'male',
      translate: 'Macho'
    },
    {
      key: 'female',
      translate: 'Fêmea'
    }
  ];

  return genders.find(element => element.key === gender).translate;
};

export const getAnimalSize = (size) => {
  const sizes = [
    {
      key: 'small',
      translate: 'Pequeno'
    },
    {
      key: 'medium',
      translate: 'Médio'
    },
    {
      key: 'bigger',
      translate: 'Grande'
    }
  ];

  return sizes.find(element => element.key === size).translate;
};

export const uploadFile = async (upload) => {
  let url = '';
  // eslint-disable-next-line new-cap
  const base64Data = new Buffer.from(
    upload.preview.uri.replace(/^data:image\/\w+;base64,/, ''),
    'base64'
  );

  const params = {
    Bucket: S3_BUCKET,
    Key: upload.image.name,
    Body: base64Data,
    ACL: 'public-read',
    ContentEncoding: 'base64',
    ContentType: upload.image.type
  };

  const s3 = new AWS.S3({
    accessKeyId: S3_KEY,
    secretAccessKey: S3_SECRET,
    region: S3_REGION
  });

  try {
    const { Location } = await s3.upload(params).promise();
    url = Location;
  } catch (err) {
    throw err;
  }

  return url;
};
