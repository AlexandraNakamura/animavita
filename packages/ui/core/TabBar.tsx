import React, {useEffect, useRef, useState} from 'react';
import styled, {css} from 'styled-components/native';
import {LayoutChangeEvent, StyleSheet} from 'react-native';
import Animated, {Easing} from 'react-native-reanimated';

import {PossibleThemes, px2ddp, StyledTheme, useTheme} from '@animavita/theme';

import Row from '../layout/Row';
import Space from '../layout/Space';

import Typography from './Typography';

const {timing} = Animated;

const Wrapper = styled.View`
  width: 100%;
`;
const ClickableArea = styled.TouchableOpacity``;
const LineWrapper = styled.View`
  position: relative;
`;
const BaseLine = styled.View<{themeName: PossibleThemes}>`
  position: absolute;
  background-color: ${({theme, themeName}) => (themeName === 'light' ? theme.black : theme.white)};
  height: ${StyleSheet.hairlineWidth}px;
  width: 100%;
  opacity: 0.4;
`;

function useWidthOfSelectedItem(selectedIndex: number, itemsLength: number) {
  const [widthOfSelected, setWidthOfSelected] = useState(0);

  const widthOfEachItem = useRef<number[]>([...Array(itemsLength).keys()]);

  useEffect(() => {
    setWidthOfSelected(widthOfEachItem.current[selectedIndex]);
  }, [selectedIndex]);

  const updateSelectedWidth = (event: LayoutChangeEvent, index: number) => {
    const {width} = event.nativeEvent.layout;
    widthOfEachItem.current[index] = width;
    if (index === selectedIndex) setWidthOfSelected(width);
  };
  return {widthOfSelected, updateSelectedWidth};
}

function useAnimatedLineWidth(widthOfSelected: number) {
  const widthValue = useRef<number>(0);
  const width = new Animated.Value<number>(widthValue.current);
  const [hasInteracted, setHasInteracted] = useState(false);

  function interact() {
    if (!hasInteracted) setHasInteracted(true);
  }

  useEffect(() => {
    const toValue = widthOfSelected;

    timing(width, {
      duration: 500,
      toValue,
      easing: Easing.inOut(Easing.ease),
    }).start();

    widthValue.current = toValue;
  }, [widthOfSelected]);

  return {width: hasInteracted ? width : widthOfSelected, interact};
}

interface Item {
  key: string;
  displayName: string;
}

interface TabBarProps {
  items: Item[];
  onPress: (key: string) => void;
  indexOfStartSelected?: number;
}

const TabBar: React.FC<TabBarProps> = ({items, onPress, indexOfStartSelected}) => {
  const {themeName} = useTheme();

  const [selectedIndex, setSelectedIndex] = useState(indexOfStartSelected || 0);
  const {widthOfSelected, updateSelectedWidth} = useWidthOfSelectedItem(selectedIndex, items.length);

  const {width, interact} = useAnimatedLineWidth(widthOfSelected);

  const handleItemPress = (key: string, index: number) => {
    interact();
    onPress(key);
    setSelectedIndex(index);
  };

  return (
    <Wrapper>
      <LineWrapper>
        <Animated.View
          style={{
            position: 'absolute',
            backgroundColor: themeName === 'light' ? StyledTheme.black : StyledTheme.white,
            height: StyleSheet.hairlineWidth,
            width,
          }}
        />
        <BaseLine themeName={themeName} />
      </LineWrapper>
      <Space height={px2ddp(2)} />
      <Row
        css={css`
          width: 100%;
          justify-content: space-between;
        `}>
        {items.map((item, index) => (
          <ClickableArea key={index} onPress={() => handleItemPress(item.key, index)}>
            <Typography
              onLayout={event => updateSelectedWidth(event, index)}
              variant="body"
              css={css`
                opacity: ${index === selectedIndex ? 1 : 0.4};
              `}>
              {item.displayName.toUpperCase()}
            </Typography>
          </ClickableArea>
        ))}
      </Row>
    </Wrapper>
  );
};

export default TabBar;
