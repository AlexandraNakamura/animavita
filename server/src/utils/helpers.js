export const getAnimalGender = (gender) => {
  const genders = [
    {
      key: 'male',
      trans: 'Macho'
    },
    {
      key: 'female',
      trans: 'Fêmea'
    }
  ];

  return genders.find(element => element.key === gender).trans;
};

export const getAnimalSize = (size) => {
  const sizes = [
    {
      key: 'small',
      trans: 'Pequeno'
    },
    {
      key: 'medium',
      trans: 'Médio'
    },
    {
      key: 'bigger',
      trans: 'Grande'
    }
  ];

  return sizes.find(element => element.key === size).trans;
};
