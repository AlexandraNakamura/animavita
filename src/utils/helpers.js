import AsyncStorage from '@react-native-community/async-storage';

export const sanitizeUsername = (name) => {
  if (name && name.length > 8) {
    return `\n ${name}`;
  }

  return name;
};

export const isEmpty = (array) => {
  if (array instanceof Array && array.length === 0) {
    return true;
  }

  return false;
};

export const getUser = async () => JSON.parse(await AsyncStorage.getItem('@animativa:user'));
