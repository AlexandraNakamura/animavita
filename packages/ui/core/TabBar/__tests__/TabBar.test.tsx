import React from 'react';
import {fireEvent} from '@testing-library/react-native';
import renderer from 'react-test-renderer';

import {Mount, ThemeContext} from '../../../tests/helpers';
import TabBar from '../index';

jest.mock('react-native-reanimated', () => require('react-native-reanimated/mock'));

const items = [
  {displayName: 'adoções', key: 'adocoes'},
  {displayName: 'favoritos', key: 'favoritos'},
  {displayName: 'solicitações', key: 'solicitacoes'},
];

const defaultProps = {
  onPress: jest.fn().mockName('onPress'),
  items,
};

const mountFactory = propOverrides => {
  const tabBarProps = {...defaultProps, ...propOverrides};

  const element = Mount(<TabBar {...tabBarProps} />);

  return element;
};

describe('TabBar', () => {
  it('calls onPress callback', () => {
    const cb = jest.fn();

    const {getByText} = mountFactory({onPress: cb});

    fireEvent.press(getByText('FAVORITOS'));

    expect(cb).toHaveBeenCalledWith('favoritos');
  });

  it('highlights selected tab', () => {
    const {getByText} = mountFactory({indexOfStartSelected: 1});

    expect(getByText('FAVORITOS')).toHaveStyle({opacity: 1});
  });

  it('renders correctly', () => {
    const tree = renderer
      .create(ThemeContext(<TabBar items={defaultProps.items} onPress={defaultProps.onPress} />))
      .toJSON();
    expect(tree).toMatchSnapshot();
  });
});
