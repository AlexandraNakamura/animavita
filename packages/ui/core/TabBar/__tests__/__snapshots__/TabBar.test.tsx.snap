// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TabBar renders correctly 1`] = `
<View
  style={
    Array [
      Object {
        "width": "100%",
      },
    ]
  }
>
  <View
    style={
      Array [
        Object {
          "position": "relative",
        },
      ]
    }
  >
    <View
      style={
        Object {
          "backgroundColor": "#353434",
          "height": 0.5,
          "left": Object {
            "setValue": [Function],
          },
          "position": "absolute",
          "width": 0,
        }
      }
    />
    <View
      onLayout={[Function]}
      style={
        Array [
          Object {
            "backgroundColor": "#353434",
            "height": 0.5,
            "opacity": 0.4,
            "position": "absolute",
            "width": "100%",
          },
        ]
      }
      themeName="light"
    />
  </View>
  <View
    height={4}
    style={
      Array [
        Object {
          "height": 4,
        },
      ]
    }
  />
  <View
    css={
      Array [
        "
          width: 100%;
          justify-content: space-between;
        ",
      ]
    }
    style={
      Array [
        Object {
          "alignItems": "center",
          "flexDirection": "row",
          "justifyContent": "space-between",
          "width": "100%",
        },
      ]
    }
  >
    <TouchableOpacity
      activeOpacity={0.2}
      onPress={[Function]}
      style={
        Array [
          Object {},
        ]
      }
    >
      <Text
        css={
          Array [
            "
                opacity: ",
            "1",
            ";
              ",
          ]
        }
        onLayout={[Function]}
        style={
          Array [
            Object {
              "color": "#353434",
            },
            Object {
              "fontSize": 26.5,
              "opacity": 1,
            },
          ]
        }
        themeName="light"
      >
        ADOÇÕES
      </Text>
    </TouchableOpacity>
    <TouchableOpacity
      activeOpacity={0.2}
      onPress={[Function]}
      style={
        Array [
          Object {},
        ]
      }
    >
      <Text
        css={
          Array [
            "
                opacity: ",
            "0.4",
            ";
              ",
          ]
        }
        onLayout={[Function]}
        style={
          Array [
            Object {
              "color": "#353434",
            },
            Object {
              "fontSize": 26.5,
              "opacity": 0.4,
            },
          ]
        }
        themeName="light"
      >
        FAVORITOS
      </Text>
    </TouchableOpacity>
    <TouchableOpacity
      activeOpacity={0.2}
      onPress={[Function]}
      style={
        Array [
          Object {},
        ]
      }
    >
      <Text
        css={
          Array [
            "
                opacity: ",
            "0.4",
            ";
              ",
          ]
        }
        onLayout={[Function]}
        style={
          Array [
            Object {
              "color": "#353434",
            },
            Object {
              "fontSize": 26.5,
              "opacity": 0.4,
            },
          ]
        }
        themeName="light"
      >
        SOLICITAÇÕES
      </Text>
    </TouchableOpacity>
  </View>
</View>
`;
